### ISSUE TITLE
Support open source nfc hardware tokens

### ISSUE THREAD

**Issue by J-Jamet** [Kunzisoft/KeePassDX#304]
like https://www.kickstarter.com/projects/conorpatrick/solo-the-first-open-source-fido2-security-key-usb?lang=de (email request)

**Comment by Iolaum** [comment]
I have solo keys. Feel free to contact me if you need testers for this feature.

**Comment by rugk** [comment]
Well… yeah, good idea. Given the cited key (SoloKey) now has [the vs campaign running](https://www.kickstarter.com/projects/conorpatrick/solo-v2-safety-net-against-phishing), here are some more details on how that would have to be implemented:

In contrast to #8, which uses HMAC-SHA1, which is a proprietary extension from Yubikey, the thing better to be done is use `hmac-secret`, which **any FIDO U2F/WebAuthn token supports**.
According to [some comment by the creators of the token](https://github.com/solokeys/kickstarter2021/discussions/23#discussioncomment-312837) here is how that should be implemented:
> We still think the way to do this is to use the `hmac-secret` extension that is part of the actual standard (it is in essence HMAC-SHA256 challenge-response). In that sense, the answer is, yes, at launch (and Solo V1 has it too, just like every fully certified FIDO2 dongle).
> If the question is whether we will implement Yubico's proprietary app (in essence, HMAC-SHA1 challenge-response), the answer is no, we do not plan to do so. This sentiment is shared by other vendors such as Trezor - if OSS such as KeePassXC and LUKS moves on to an actual standard instead of one company's private sauce, everyone wins.

"Upstream" **KeePassXC** issue for that is: https://github.com/keepassxreboot/keepassxc/issues/3560
KeeWeb: https://github.com/keeweb/keeweb/issues/1681

**Comment by rugk** [comment]
Also I'd propose to rename this issue to a more accurate title like "Add support for hmac-secret FIDO2 extension".

**Comment by 4n0nym0u53** [comment]
I would like to add, that the use of `hmac-secret` for FIDO2 would defeat the main purpose of implementing support for hardware keys (at least for me). IMO, the main purpose of this would be to have compatibility with KeePassXC for desktop, which uses HMAC-SHA1 for YubiKeys (and OnlyKeys too). This missing functionality is in fact the only reason for why I haven't switched to KeePassDX. I currently use Keepass2Android, which uses the [ykDroid](https://github.com/pp3345/ykDroid) driver app (available on F-Droid) to achieve support for HMAC-SHA1, but would like to switch to KeePassDX in the near future due to better looking UI.

Implementing both could also work. I think that HMAC-SHA1 implementation is important in order to ensure cross-compatibility with other KeePass implemtations (such as KeePassXC for desktop). I don't know of any desktop KeePass programs supporting `hmac-secret`, but if you know any I'll be more than happy to check them out.

**Comment by rugk** [comment]
> IMO, the main purpose of this would be to have compatibility with other KeePass programs, such as KeePassXC for desktop, which uses HMAC-SHA1 for YubiKeys (and OnlyKeys too).

THat’s a problem of KeePassXC, actually and tracked in https://github.com/keepassxreboot/keepassxc/issues/3560 as said. As argued, it makes no sense to only support YubiKeys, when you can support a standard and thus all keys.

The feature request you are probably looking for is thus https://github.com/Kunzisoft/KeePassDX/issues/8, which is about YubiKey.

**Comment by 4n0nym0u53** [comment]
@rugk as I understand it, `hmac-secret` makes use of the normal FIDO2 flow used for authentication in e.g. browsers (correct me if I'm wrong). In that case, please note that degoogled phones don't have native support for FIDO/FIDO2, as it a part of Google Play Services, so KeePassDX should have it's own implementation instead of relying on Android's (or rather, Google's) implementation.

AFAIK, FIDO is only supported in Chrome, so KeePassDX would need its own implementation anyway. Also, I think that omitting the use of Google Play Services is important for a FOSS project like this.

**Comment by J-Jamet** [comment]
Here is an alternative FIDO2 SDK in GPL3 that does not require the play service, it may be a good starting point to study the feasibility.
https://github.com/cotechde/hwsecurity

**Comment by J-Jamet** [comment]
> @rugk I would like to add, that the use of hmac-secret for FIDO2 would defeat the main purpose of implementing support for hardware keys (at least for me). IMO, the main purpose of this would be to have compatibility with other KeePass programs, such as KeePassXC for desktop, which uses HMAC-SHA1 for YubiKeys (and OnlyKeys too). This missing functionality is in fact the only reason for why I haven't switched to KeePassDX. I currently use Keepass2Android, which uses the ykDroid driver app (available on F-Droid) to achieve support for HMAC-SHA1, but I would like to switch to KeePassDX in the near future due to better looking UI.

The goal would be to make as many protocols and physical keys as possible compatible with KeePass apps. Historically, Yubikey was the first to be implemented (and not in a homogeneous way), I promote backward compatibility but we might as well favor open solutions for new implementations.
There is nothing to stop other KeePass applications from evolving by making contributions and pull requests to make FIDO2 supported.

**Comment by rugk** [comment]
Note that on Android by default all apps/browsers seem to depend on Google Play Services (yet again :roll_eyes:) for FISO2/WebAuthn to work/use that implementation. A feature request for implementation at microG (https://github.com/microg/GmsCore/issues/849) is still open.

**Comment by BryanJacobs** [comment]
I've raised a pull request again hwsecurity to add support for hmac-secret. No response yet from the maintainers, but the GPLv3 license of that library would permit bundling a fork of it into the hardware key driver code.

I would recommend against implementing a solution that directly uses the FIDO2 hmac-secret challenge response as part of the keying material for the password database. If you do that, users will have one hardware token whose loss would mean being unable to open their database. To me that's not an acceptable state of affairs...

Instead, the database should store an encryption of the primary keying material part with n-many different hmac-secret challenge responses. That way one of n-many different FIDO2 devices could be used to unlock the database.

**Comment by J-Jamet** [comment]
I agree with that, the problem would be where to store the n-many. The Keeweb WIP explains the concept well but will putting this info in the database header pose a security problem?
Some KeePass plugins used the method of creating a separate file for the increments but I don't really like this method because the goal of KeePass is to have only one file to manage.

I will remove the option "FIDO2 secret" in the selection of the hardware key of KeePassDX because the functionality still needs to be discussed further and standardized.

**Comment by ale5000-git** [comment]
The latest release of microG have FIDO/FIDO2/WebAuthn support, while currently is only for browsers at least we already are some steps ahead.
An app can include only `play-services-fido` to work on devices with microG, or it can include both `play-services-fido` AND `play-services-fido-core` to work with and without microG.

**Comment by ashleysommer** [comment]
Good news on this topic.
The implementation of FIDO2 "hmac-secret" in KeePassXC is going forward. (I am the volunteer making the implementation based on libfido2). See the aforementioned KeepassXC issue thread, and [this discussion topic](https://github.com/keepassxreboot/keepassxc/discussions/9506) for further details.

As a user of KeePassDX on Android, it would be great to see this also implemented here in a compatible manner, so that a keepass database created in KeePassXC secured with a hmac-secret device could be opened in KeepassDX with the same key.

The challenges as discussed above is the primarily which FIDO2 library to pick. The official android FIDO2 api does support "hmac-secret" but is tied to Google Play Services, which I believe we want to avoid.

The mentioned alternative library https://github.com/cotechde/hwsecurity is out of date, unmaintained, and does not support "hmac-secret".

The latest release of microG has its own FIDO2 implementation, as mentioned above, and while its only intended for browsers it may be able to work directly with this app. Does KeePassDX use microG elsewhere already?

There is a fully independent complete FIDO2 implementation created by the NitroKey team, but I believe it is proprietary, it requires a payment or subscription: https://www.nitrokey.com/products/android-fido2-sdk

**Comment by J-Jamet** [comment]
KeePassDX doesn't use microG, but there's nothing to stop extracting the source code and turning it into a driver app.

**Comment by ashleysommer** [comment]
I just noticed the existing KeePassDX Security Key Hardware Driver https://gitlab.com/kunzisoft/android-hardware-key-driver already mentions planned support for "hmac-secret" extension (For the solo key, but I don't see why it wouldn't work on other keys).

**Comment by samsapti** [comment]
@ashleysommer the one from Nitrokey is the same as the [Hardware Security SDK](/cotechde/hwsecurity). It's dual-licensed as GPLv3 and a proprietary license.

**Comment by ashleysommer** [comment]
@samsapti Thanks for pointing that out. Looks like that library doesn't yet support "hmac-secret" extension, but there is an open pull request (https://github.com/cotechde/hwsecurity/pull/5) to add support.

**Comment by BryanJacobs** [comment]
I opened that pull request, and since nobody has responded in the months since, I think it's safe to say that hwsecurity is indeed unmaintained.

**Comment by ashleysommer** [comment]
Ah right, yep I didn't to see the connection until now, with out comment from earlier in this thread.

**Comment by lazytownfan** [comment]
I just wanted to bring up that recently KeePassXC added support for the Nitrokey 3, but using `challenge-response`.

Does this in any way help support for SoloKeys?

**Comment by ashleysommer** [comment]
@lazytownfan 
> recently KeePassXC https://github.com/keepassxreboot/keepassxc/pull/9631 support for the Nitrokey 3

> Does this in any way help support for SoloKeys?

No, this is unrelated to the FIDO2 efforts that will enable support for SoloKeys and other FIDO2 keys.

> does this mean the Nitrokey 3 will be able to work with the latest version of KeePassDX, just like the YubiKey

Yes. That feature is merged, and be part of the v2.8.0 release of KeePassXC, I don't know when the release date for that will be.

**Comment by rugk** [comment]
@J-Jamet AFAIk vendors are usually quite helpful on supporting software devs with a hardware sample, so you may just contact @nitrokey to get a key.

**Comment by jans23** [comment]
@J-Jamet For development purposes we can provide you a free Nitrokey 3. Please DM me or send us an email to arrange that.

**Comment by BryanJacobs** [comment]
I am in the process of writing a suitable FIDO2 Platform implementation that could be used for this feature: https://github.com/BryanJacobs/FIDOk/ .

It's not ready yet: although technically things work with USB or NFC authenticators on Android, and it supports HMAC-Secret just fine, the documentation, error handling, and test coverage aren't up to snuff.

But soon it will be able to serve as an alternative to both the hwsecurity code and to Google's own half-baked FIDO implementation.

**Comment by BryanJacobs** [comment]
I've raised https://gitlab.com/kunzisoft/android-hardware-key-driver/-/merge_requests/12 to make supporting hmac-secret here possible.

**Comment by CoelacanthusHex** [comment]
> Note that on Android by default all apps/browsers seem to depend on Google Play Services (yet again 🙄) for FIDO2/WebAuthn to work/use that implementation. A feature request for implementation at microG ([microg/GmsCore#849](https://github.com/microg/GmsCore/issues/849)

microG has implemented FIDO feature since [v0.2.25.223616](https://github.com/microg/GmsCore/releases/tag/v0.2.25.223616) (2022-10-21) and implemented FIDO PIN support since [v0.3.3.240913](https://github.com/microg/GmsCore/releases/tag/v0.3.3.240913) (2024-09-28).

Fennec (a libre build of Firefox) has used microG as a replacement for the GMS library to implement the FIDO feature. It can be done by simply replacing `com.google.android.gms:play-services-fido` with `org.microg.gms:play-services-fido`. See https://gitlab.com/relan/fennecbuild/-/blob/master/gecko-liberate.patch. There is more work in Fennec's patch because Firefox has some code like `if (BuildConfig.MOZILLA_OFFICIAL)`, we don't need it.

**Comment by fuzzblob** [comment]
What about setups that wish to avoid connecting to google altogether and don't have GSF or microG installed at all?

**Comment by CoelacanthusHex** [comment]
> What about setups that wish to avoid connecting to google altogether and don't have GSF or microG installed at all?

Actually, microG's FIDO part can be used as a standalone full implementation library with a GMS-like API. Fennec considered this way before.

Also, the FIDO part of the microG service is unrelated to other parts connecting to the Google server. You can disable the network permission of microG and use its FIDO or other networkless features. I see some users use microG this way.

**Comment by scrdcow** [comment]
Hi! How is this going? I see it in todo. but not started. I'm not a dev myself so can't help much on that regard :-/ Are there anything I can do to help. I have a Nitrokey 3C NFC so I can test code atleast.

**Comment by J-Jamet** [comment]
I understand that many people are waiting for this feature. I need to refactor the Key Driver code and review what @BryanJacobs has done, thank you very much it's really appreciated.  👍 This takes time because I have to prioritize. Version 4.3.0 already contains many features, so I am postponing it to version 4.4.0, but it will of course be reviewed.

**Comment by J-Jamet** [comment]
This issue is progressing well. I am currently refactoring the transport code in KeePassDX that retrieves the seed and credentialId from the database to request the challenge from a Yubikey HMAC-SHA1 or a FIDO2 hmac-secret. The thing is, we need to agree on the format to put in the database header. I haven't read the discussion https://github.com/keepassxreboot/keepassxc/discussions/9506 on the implementation of the XML blob in KeePasXC yet, but I'll look into it.

@BryanJacobs I did a quick first pass on the Fidok code in the driver application, and I'm going to merge it into the develop branch so that I can then migrate to Flow and Compose for a better architecture. Even if there are changes to be made, it will be on top of your work.

**Comment by J-Jamet** [comment]
So the basic workflow that allows to send the salt and credentialId from the header works with the current code in feature/fido2 branch using the new version 0.2.0 of Key-Driver.

Remaining tasks: 
 1. Security checks to prevent replay and add signature verification.
 2. Modifying the basic workflow to manage an encrypted blob in the database header so that key backup can be created in case of hardware key loss. Again, thx @BryanJacobs, you explained that very well in your comment on KeePassXC.. https://github.com/keepassxreboot/keepassxc/discussions/9506#discussioncomment-6137689 I'll start with your guidelines for now and see what constraints arise as we go along. I was thinking of creating the “Master secret part / backup key” from the BIP-39 mnemonic keywords, so that they could be stored on a piece of paper.
 3. Allow the addition of a new master password factor enabling key backup in KeePassDX. Ultimately, it would be advantageous to have a modular system allowing N factors to be defined as the master password in order to combine several factors of the same type, but a new UI must be considered. This issue will be addressed in another issue.
 4. At the Key-Driver level, the application requires the FIDO key PIN for each CREATION and GET actions, but to create a database or change the keepass master password, both actions are required, so the PIN is called several times in quick succession. It should be possible to store it in the keystore so that it can be accessed using biometrics. I will open a new issue for this.
 5. Unit tests and comprehensive doc explaining how it works, it is much more complex than Yubikey's HMAC-SAH1 challenge response.

With the amount of work that still needs to be done, the feature will be included in a version higher than 4.4, but developed in parallel so that it can be tested as it progresses.

**Comment by BryanJacobs** [comment]
> So the basic workflow that allows to send the salt and credentialId from the header works with the current code in feature/fido2 branch using the new version 0.2.0 of Key-Driver.
> 
> Remaining tasks:
> 
>     1. Security checks to prevent replay and add signature verification.

Note the code in https://gitlab.com/kunzisoft/android-hardware-key-driver/-/merge_requests/12 allows performing signature verification. I'm not sure replay prevention beyond what normal FIDO2 authenticate calls do is very useful for this case though.

>     2. Modifying the basic workflow to manage an encrypted blob in the database header so that key backup can be created in case of hardware key loss. Again, thx [@BryanJacobs](https://github.com/BryanJacobs), you explained that very well in your comment on KeePassXC.. [Implementation of FIDO2 hmac-secret extension in KeePassXC keepassxreboot/keepassxc#9506 (reply in thread)](https://github.com/keepassxreboot/keepassxc/discussions/9506#discussioncomment-6137689) I'll start with your guidelines for now and see what constraints arise as we go along. I was thinking of creating the “Master secret part / backup key” from the BIP-39 mnemonic keywords, so that they could be stored on a piece of paper.

Glad you found it useful. Yes, you can make the master secret part however you want.

Note that due to comments from the keepassXC maintainer I am in the process of changing the header format to an (exactly equivalent) JSON format instead of XML. Otherwise identical.

>     3. Allow the addition of a new master password factor enabling key backup in KeePassDX. Ultimately, it would be advantageous to have a modular system allowing N factors to be defined as the master password in order to combine several factors of the same type, but a new UI must be considered. This issue will be addressed in another issue.

Wise.

>     4. At the Key-Driver level, the application requires the FIDO key PIN for each CREATION and GET actions, but to create a database or change the keepass master password, both actions are required, so the PIN is called several times in quick succession. It should be possible to store it in the keystore so that it can be accessed using biometrics. I will open a new issue for this.

You should NOT need to call a function more than one time to change the password. When opening the database, use both HMAC salts (salt1 and salt2). Salt1 should be the value to open the database now (previous key). Salt2 should be prospective/tentative key to use on rotation.

If the user doesn't decide to change the master password, throw away the result you got with salt2. If the user does... you don't need to call the authenticator again, because you already got back hmac(salt2).

>     5. Unit tests and comprehensive doc explaining how it works, it is much more complex than Yubikey's HMAC-SAH1 challenge response.

Agreed.

> 
> With the amount of work that still needs to be done, the feature will be included in a version higher than 4.4, but developed in parallel so that it can be tested as it progresses.

That's fair.

**Comment by J-Jamet** [comment]
> JSON format instead of XML

I haven't read everything in detail, but isn't it more logical to maintain consistency in formats rather than adding new ones? I will respond in the KeePassDX discussion once I reread and understand the reasons.

> You should NOT need to call a function more than one time to change the password. When opening the database, use both HMAC salts (salt1 and salt2). Salt1 should be the value to open the database now (previous key). Salt2 should be prospective/tentative key to use on rotation.

You're right, I missed that information, I will change the workflow accordingly.

**Comment by rugk** [comment]
For reference: https://gitlab.com/kunzisoft/android-hardware-key-driver recently integrated the Kotlin lib https://github.com/BryanJacobs/FIDOk/ for this functionality (if I understand it well) in https://gitlab.com/kunzisoft/android-hardware-key-driver/-/merge_requests/12

**Comment by BryanJacobs** [comment]
You wouldn't have to add a FIDO library directly here in KeePassDX if you were to use the Intent-based API there. Instead users could install an app implementing that intent (like android-hardware-key-driver). Very similar to the current way hmac-sha1 works.
